### PR TITLE
EVG-19951: Add bootstrap methods to host settings

### PR DESCRIPTION
**Pull request by sophstad**
EVG-19951

### Description

- Add non-Legacy SSH bootstrap config options
- Add missing descriptions

### Screenshots

image

### Testing

- Add integration tests

### CODE REVIEW

File: src/pages/distroSettings/tabs/HostTab/getFormSchema.tsx [hunk 1]
```diff
@@ -157,12 +190,127 @@ export const getFormSchema = ({
...
+          ...(hasStaticProvider && {
+            "ui:disabled": true,
+            "ui:tooltipDescription": "Static distros are not spawnable.",
+          }),
+          "ui:description":
+            "Allow users to spawn these hosts for personal use.",
+          "ui:bold": true,
+        },
+        isVirtualWorkStation: {
+          "ui:description":
```
Comment (minnakt) [inline src/pages/distroSettings/tabs/HostTab/getFormSchema.tsx:220]: it seems this should also be disabled if the arch is [not Linux](https://github.com/evergreen-ci/evergreen/blob/dfb07d9e59ef9db396fdb2e55a3660259b430667/service/templates/distros.html#L1065) 😞

File: src/pages/distroSettings/tabs/HostTab/getFormSchema.tsx [hunk 2]
```diff
@@ -157,12 +190,127 @@ export const getFormSchema = ({
...
+            "ui:description":
+              "Max number of cgroup tasks (threads). Set -1 for unlimited.",
+          },
+          lockedMemory: {
+            "ui:description":
+              "Max size (kB) that can be locked into memory. Set -1 for unlimited.",
+          },
+          virtualMemory: {
+            "ui:description":
+              "Max size (kB) of available virtual memory. Set -1 for unlimited.",
```
Comment (minnakt) [inline src/pages/distroSettings/tabs/HostTab/getFormSchema.tsx:278]:
```suggestion
          lockedMemoryKb: {
            "ui:description":
              "Max size (kB) that can be locked into memory. Set -1 for unlimited.",
          },
          virtualMemoryKb: {
            "ui:description":
              "Max size (kB) of available virtual memory. Set -1 for unlimited.",
```

File: src/pages/distroSettings/tabs/HostTab/getFormSchema.tsx [hunk 3]
```diff
@@ -157,12 +190,127 @@ export const getFormSchema = ({
         arch: {
```
Comment (minnakt) [inline src/pages/distroSettings/tabs/HostTab/getFormSchema.tsx:190]: I kind of notice that it gets harder and harder to cross-reference the schema with the uiSchema... Maybe we should try to put them in a consolidated object like `{ schema: <stuff here>, uiSchema: <stuff here> }` (and maybe I should do this in my provider PR too 😅)

File: src/pages/distroSettings/tabs/HostTab/getFormSchema.tsx [hunk 4]
```diff
@@ -1,24 +1,34 @@
...
+  windowsArchitectures,
+} from "./constants";
 
 type FormSchemaParams = {
+  architecture: Arch;
   provider: Provider;
   sshKeys: SshKey[];
 };
 
 export const getFormSchema = ({
```
Comment (minnakt) [inline src/pages/distroSettings/tabs/HostTab/getFormSchema.tsx:30]: Should we also try surfacing this error through the `validate()` function?

Screenshot 2023-09-22 at 12 34 47 PM

**Review by minnakt — COMMENTED** [review]
great job bringing together all the different parts :)

File: cypress/integration/distroSettings/host_section.ts [hunk 5]
```diff
@@ -57,4 +57,55 @@ describe("host section", () => {
...
+      cy.getInputByLabel("Locked Memory").type("128");
+      cy.getInputByLabel("Virtual Memory").type("256");
+
+      cy.contains("button", "Add variable").click();
+      cy.getInputByLabel("Key").type("my-key");
+      cy.getInputByLabel("Value").type("my-value");
+
+      cy.contains("button", "Add script").click();
+      cy.getInputByLabel(/^Path$/).type("/path/to/precondition/script");
+      cy.getInputByLabel(/^Script$/).type("script contents here");
```
[the remark this answers is not in the document]
Comment (sophstad) [reply cypress/integration/distroSettings/host_section.ts:104]: These inputs refer to the precondition script fields; the regex is actually used to make them _more_ specific so that we target the "Script" input instead of "Setup script" 
image

Using a regex is probably the best way to achieve this since `getInputByLabel` uses `.contains()`

File: src/pages/distroSettings/tabs/HostTab/getFormSchema.tsx [hunk 6]
```diff
@@ -53,95 +63,118 @@ export const getFormSchema = ({
...
+            arch: {
+              oneOf: [
+                {
+                  properties: {
+                    arch: { enum: windowsArchitectures },
+                    rootDir: {
+                      type: "string" as "string",
+                      title: "Root Directory",
+                    },
+                  },
```
[the remark this answers is not in the document]
Comment (sophstad) [reply src/pages/distroSettings/tabs/HostTab/getFormSchema.tsx:91]: Good catch, I moved it to the bottom of the section since that is significantly easier than putting it elsewhere 😅 let me know if it seems reasonable...
image

File: cypress/integration/distroSettings/host_section.ts [hunk 7]
```diff
@@ -57,4 +57,55 @@ describe("host section", () => {
...
+    });
+
+    it("shows Windows-only fields when the architecture is updated", () => {
+      cy.contains("label", "Root Directory").should("not.exist");
+      cy.contains("label", "Service User").should("not.exist");
+
+      cy.selectLGOption("Agent Architecture", "Windows 64-bit");
+
+      cy.getInputByLabel("Root Directory").should("exist");
+      cy.getInputByLabel("Service User").should("exist");
```
[the remark this answers is not in the document]
Comment (sophstad) [reply cypress/integration/distroSettings/host_section.ts:75]: These fields are out of the viewport when they are toggled, so `visible` fails without first including a `scrollIntoView`. In my opinion this is more complicated than it is helpful so I would prefer to check existence.

File: cypress/integration/distroSettings/host_section.ts [hunk 8]
```diff
@@ -57,4 +57,55 @@ describe("host section", () => {
       cy.validateToast("success");
     });
   });
+
+  describe("using User Data bootstrap method", () => {
+    beforeEach(() => {
+      cy.visit("/distro/ubuntu1604-parent/settings/host");
```
[the remark this answers is not in the document]
Comment (sophstad) [reply cypress/integration/distroSettings/host_section.ts:63]: Agree this would be helpful, but Cypress totally chokes when working with any value from `constants/routes`. Would prefer to resolve this separately!

```
Error: Webpack Compilation Error
./src/utils/string/githubPRLinkify.tsx 10:158
Module parse failed: Unexpected token (10:158)
File was processed with these loaders:
 * ../../Library/Caches/Cypress/12.17.1/Cypress.app/Contents/Resources/app/packages/server/node_modules/ts-loader/index.js
You may need an additional loader to handle the result of these loaders.
| var styles_1 = require("components/styles");
| var githubPRLinkify = function (unlinkified) {
>     return (0, react_string_replace_1.default)(unlinkified, /(https:\/\/github.com\/[a-zA-Z0-9-]+\/[a-zA-Z0-9-]+\/pull\/\d+)/g, function (match, i) { return (<styles_1.StyledLink key={"".concat(match).concat(i)} href={match}>
|         {match}
|       </styles_1.StyledLink>); });
 @ ./src/utils/string/index.ts 9:24-52
 @ ./src/utils/index.ts
 @ ./src/constants/routes.ts
 @ ./cypress/integration/distroSettings/host_section.ts
    at Watching.handle [as handler] (/Users/sophstad/Library/Caches/Cypress/12.17.1/Cypress.app/Contents/Resources/app/packages/server/node_modules/@cypress/webpack-preprocessor/dist/index.js:212:23)
    at /Users/sophstad/Library/Caches/Cypress/12.17.1/Cypress.app/Contents/Resources/app/node_modules/webpack/lib/Watching.js:99:9
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/Users/sophstad/Library/Caches/Cypress/12.17.1/Cypress.app/Contents/Resources/app/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
    at Watching._done (/Users/sophstad/Library/Caches/Cypress/12.17.1/Cypress.app/Contents/Resources/app/node_modules/webpack/lib/Watching.js:98:28)
    at /Users/sophstad/Library/Caches/Cypress/12.17.1/Cypress.app/Contents/Resources/app/node_modules/webpack/lib/Watching.js:73:19
    at Compiler.emitRecords (/Users/sophstad/Library/Caches/Cypress/12.17.1/Cypress.app/Contents/Resources/app/node_modules/webpack/lib/Compiler.js:499:39)
    at /Users/sophstad/Library/Caches/Cypress/12.17.1/Cypress.app/Contents/Resources/app/node_modules/webpack/lib/Watching.js:54:20
    at /Users/sophstad/Library/Caches/Cypress/12.17.1/Cypress.app/Contents/Resources/app/node_modules/webpack/lib/Compiler.js:485:14
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/Users/sophstad/Library/Caches/Cypress/12.17.1/Cypress.app/Contents/Resources/app/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
    at /Users/sophstad/Library/Caches/Cypress/12.17.1/Cypress.app/Contents/Resources/app/node_modules/webpack/lib/Compiler.js:482:27
    at /Users/sophstad/Library/Caches/Cypress/12.17.1/Cypress.app/Contents/Resources/app/node_modules/neo-async/async.js:2818:7
    at done (/Users/sophstad/Library/Caches/Cypress/12.17.1/Cypress.app/Contents/Resources/app/node_modules/neo-async/async.js:3522:9)
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/Users/sophstad/Library/Caches/Cypress/12.17.1/Cypress.app/Contents/Resources/app/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
    at /Users/sophstad/Library/Caches/Cypress/12.17.1/Cypress.app/Contents/Resources/app/node_modules/webpack/lib/Compiler.js:464:33
    at /Users/sophstad/Library/Caches/Cypress/12.17.1/Cypress.app/Contents/Resources/app/packages/server/node_modules/graceful-fs/graceful-fs.js:143:16
    at /Users/sophstad/Library/Caches/Cypress/12.17.1/Cypress.app/Contents/Resources/app/packages/server/node_modules/graceful-fs/graceful-fs.js:143:16
    at /Users/sophstad/Library/Caches/Cypress/12.17.1/Cypress.app/Contents/Resources/app/packages/server/node_modules/graceful-fs/graceful-fs.js:61:14
    at FSReqCallback.oncomplete (node:fs:196:23)
```

File: src/pages/distroSettings/tabs/HostTab/getFormSchema.tsx [hunk 9]
```diff
@@ -53,95 +63,118 @@ export const getFormSchema = ({
...
               title: "Working Directory",
-            },
-            setupScript: {
-              type: "string" as "string",
-              title: "Setup Script",
+              minLength: 1,
             },
             setupAsSudo: {
               type: "boolean" as "boolean",
               title: "Run script as sudo",
```
[the remark this answers is not in the document]
Comment (sophstad) [reply src/pages/distroSettings/tabs/HostTab/getFormSchema.tsx:70]: When I was developing I thought it was never super clear that "Run script as sudo" is attached to the setup script when it was located elsewhere in the form. I added a max width to our Spruce Form elements instead; I think this is worth doing regardless because forms can get pretty wonky/hard to follow on ultrawide screens.

File: src/pages/distroSettings/tabs/HostTab/getFormSchema.tsx [hunk 10]
```diff
@@ -1,24 +1,58 @@
...
+  icecreamConfigPath,
+  icecreamSchedulerHost,
+  isVirtualWorkStation,
+  jasperBinaryDir,
+  jasperCredentialsPath,
+  lockedMemoryKb,
+  maximumHosts,
+  minimumHosts,
+  numFiles,
+  numProcesses,
```
Comment (minnakt) [inline src/pages/distroSettings/tabs/HostTab/getFormSchema.tsx:30]: For some parts of the form, I think it could help to group these fields as the section and then just import that object, so that you don't have to import a bunch of fields.

e.g. like having a `{ schema: <stuff>, uiSchema: <stuff> }` object for the whole allocation section:

Screenshot 2023-09-27 at 12 13 45 PM

File: src/pages/distroSettings/tabs/HostTab/getFormSchema.tsx [hunk 11]
```diff
@@ -1,24 +1,58 @@
...
+  version,
+  virtualMemoryKb,
+  workDir,
+} from "./schemaFields";
 
 type FormSchemaParams = {
+  architecture: Arch;
   provider: Provider;
   sshKeys: SshKey[];
 };
```
Comment (minnakt) [inline src/pages/distroSettings/tabs/HostTab/getFormSchema.tsx:53]: I'm not sure what can be done about this, but I notice I can save with an invalid configuration:

https://github.com/evergreen-ci/spruce/assets/47064971/b0c57f7a-e154-49d5-8656-ab4104e21480

If it's too difficult to address on the frontend, possibly we can just add stronger validation on the backend 😅

**Review by minnakt — APPROVED** [review]
nice work!

File: src/pages/distroSettings/tabs/HostTab/getFormSchema.tsx (hunk 11 above)
Comment (sophstad) [reply src/pages/distroSettings/tabs/HostTab/getFormSchema.tsx:53]: Good catch, would like to have airtight validation on the back end so I'll add that in EVG-20943!
